mllib/src/main/scala/org/apache/spark/ml/feature/PolynomialMapper.scala

```diff
@@ -0,0 +1,170 @@
+
+  override protected def createTransformFunc(paramMap: ParamMap): Vector => Vector = {
+    PolynomialMapper.transform(getDegree)
+  }
+
+  override protected def outputDataType: DataType = new VectorUDT()
+}
+
+object PolynomialMapper {
+  /**
+   * The number that combines k items from N items without repeat, i.e. the binomial coefficient.
+   */
+  private def binomialCoefficient(N: Int, k: Int): Int = {
+    (N - k + 1 to N).product / (1 to k).product
+  }
+
+  /**
+   * The number of monomials of a `numVariables` vector after expanding at a specific polynomial
+   * degree `degree`.
+   */
+  private def numMonomials(degree: Int, numVariables: Int): Int = {
+    binomialCoefficient(numVariables + degree - 1, degree)
+  }
+
+  /**
+   * The number of monomials of a `numVariables` vector after expanding from polynomial degree 1 to
+   * polynomial degree `degree`.
+   */
+  private def numExpandedDims(degree: Int, numVariables: Int): Int = {
+    binomialCoefficient(numVariables + degree, numVariables) - 1
+  }
+
+  /**
+   * Given a pre-built array of Double, fill it with expanded monomials until a given polynomial
+   * degree.
+   * @param values the array of Double, which represents a dense vector.
+   * @param prevStart the start offset of elements that filled in the last function call.
+   * @param prevLen the length of elements that filled in the last function.
+   * @param currDegree the current degree that we want to expand.
+   * @param finalDegree the final expected degree that we want to expand.
+   * @param nVariables number of variables in the original feature vector.
+   */
+  @tailrec
+  private def fillDenseVector(values: Array[Double], prevStart: Int, prevLen: Int, currDegree: Int,
+        finalDegree: Int, nVariables: Int): Unit = {
+
+    if (currDegree > finalDegree) {
+      return
+    }
+
+    val currExpandedVecFrom = prevStart + prevLen
+    val currExpandedVecLen = numMonomials(currDegree, nVariables)
+
+    var leftIndex = 0
+    var currIndex = currExpandedVecFrom
+
+    while (leftIndex < nVariables) {
+      val numToKeep = numMonomials(currDegree - 1, nVariables - leftIndex)
+      val prevVecStartIndex = prevStart + prevLen - numToKeep
+
+      var rightIndex = 0
+      while (rightIndex < numToKeep) {
+        values(currIndex) =
+          values(leftIndex) * values(prevVecStartIndex + rightIndex)
+        currIndex += 1
+        rightIndex += 1
+      }
+
+      leftIndex += 1
+    }
+
+    fillDenseVector(values, currExpandedVecFrom, currExpandedVecLen, currDegree + 1, finalDegree,
+      nVariables)
+  }
+
+  /**
+   * For polynomial expanding a `SparseVector`, we treat it as a dense vector and call
+   * `fillDenseVector` to fill in the `values` of `SparseVector`. For its `indices` part, we encode
+   * the indices from `nVariables` one by one, because we do not care of the real indices.
+   */
+  private def fillPseudoSparseVectorIndices(indices: Array[Int], startFrom: Int, startWith: Int) = {
```

Comment:
@mengxr I notice that it is not a feasible solution. We have to compute the real indices of the sparse vector. I'll fix it ASAP.